examples/rhai-surrogate-cache-key/src/rhai_surrogate_cache_key.rhai

```diff
@@ -0,0 +1,68 @@
+// This example illustrates how to interact with responses to
+// modify the data returned.
+
+// At the router_service stage, register callbacks for processing 
+// responses. We are using a closure here, but it could be a
+// function if we chose (see execution_service).
+fn router_service(service) {
+    let map_callback = |response| {
+        try {
+            // Extract our surrogate key from context
+            // If there is no surrogate key in our context,
+            // there is no work to do.
+            if response.context["surrogate-cache-key"] != () {
+                let header_value = "";
+                let sep = "";
+                for value in response.context["surrogate-cache-key"].values() {
+                    header_value += sep;
+                    header_value += value;
+                    sep = ",";
+                }
+                response.headers["surrogate-cache-key"] = header_value;
+            }
+        }
+        catch(err)
+        {
+            // log any errors
+            error_log(`surrogate-cache-key not created: ${err}`);
+        }
+    };
+    // Map our response using our closure
+    service.map_response(map_callback);
+}
+
+// At the subgraph_service stage, register callbacks for processing
+// responses.
+// We are going to use two closures here:
+//  - map_callback to map and manipulate the service response
+//  - resolver to provide an upserting function for the context
+//    upsert.
+// If we can't find headers or a cache-key, we don't want to fail,
+// so catch the error and log it out.
+fn subgraph_service(service, subgraph) {
+    let map_callback = |response| {
+        try {
+            // IMPORTANT: Take a copy of the cache-key here
+            // to avoid deadlock in the closure.
```

Comment:
maybe the documentation should have a section about that deadlock risk

Reply:
That's a general `rhai` risk rather than anything specific about how we use it. I will think about the best way to incorporate something that [references this](https://rhai.rs/book/language/fn-closure.html?highlight=deadlock#data-races-in-sync-builds-can-become-deadlocks).

Reply:
I've added a section to the docs: 3f08c90afd0280f728b1239b95a0cbdc24e26ef0

Reply:
sure